### PR TITLE
feat: add parallel processing to SimpleDirectoryReader 

### PR DESCRIPTION
**Pull request by ghost**
Second part of #808 

Reworked everything, a lot cleaner and simpler now and implemented straight into SimpleDirectoryReader.

Did not overload LlamaParse when testing this time.
The data in the example is now just the brk-2022.pdf split in 6 parts.

### CODE REVIEW

File: packages/core/src/readers/SimpleDirectoryReader.edge.ts [hunk 1]
```diff
@@ -58,7 +65,37 @@ export class SimpleDirectoryReader implements BaseReader {
...
     for await (const filePath of walk(fs, directoryPath)) {
+      filePathQueue.push(filePath);
+    }
+
+    const workerPromises = Array.from({ length: numWorkers }, () =>
+      this.processFiles(filePathQueue, fs, defaultReader, fileExtToReader),
+    );
+
+    const results = await Promise.all(workerPromises);
+    docs.push(...results.flat());
```
Comment (himself65) [inline packages/core/src/readers/SimpleDirectoryReader.edge.ts:79]:
```suggestion
    docs.push(...results.flat());
```

use `Array.prototype.concat`, push will throw error when results larger than 1000 items

**Review by himself65 — CHANGES_REQUESTED** [review]

File: packages/core/src/readers/SimpleDirectoryReader.edge.ts [hunk 2]
```diff
@@ -46,19 +47,53 @@ export class SimpleDirectoryReader implements BaseReader {
...
     }
 
-    const docs: Document[] = [];
+    const filePathQueue: string[] = [];
+
     for await (const filePath of walk(fs, directoryPath)) {
+      filePathQueue.push(filePath);
+    }
+
+    const workerPromises = Array.from({ length: numWorkers }, () =>
```
Comment (marcusschiesser) [inline packages/core/src/readers/SimpleDirectoryReader.edge.ts:73]: is this handling queue.length > numWorkers? maybe using something like p-limit helps?

File: packages/core/src/readers/SimpleDirectoryReader.edge.ts [hunk 3]
```diff
@@ -70,10 +105,10 @@ export class SimpleDirectoryReader implements BaseReader {
 
         let reader: BaseReader;
 
-        if (fileExtToReader && fileExt in fileExtToReader) {
-          reader = fileExtToReader[fileExt];
-        } else if (defaultReader != null) {
+        if (defaultReader) {
```
[the remark this answers is not in the document]
Comment (InsightByAI) [reply packages/core/src/readers/SimpleDirectoryReader.edge.ts:108]: Oh, misinterpreted the code then.
In that case, suggestion: add a new param to overwrite all `fileExtToReader` if defined.

Could become more important with the addition of more external readers that are capable of reading multiple formats.

File: packages/core/src/readers/SimpleDirectoryReader.edge.ts (hunk 2 above)
Comment (InsightByAI) [reply packages/core/src/readers/SimpleDirectoryReader.edge.ts:73]: Yes it is, I thought about using p-limit, I just always try to avoid adding packages if possible. But will check it out and test around. Thanks for the review!

File: packages/core/src/readers/SimpleDirectoryReader.edge.ts [hunk 4]
```diff
@@ -48,67 +56,99 @@ export class SimpleDirectoryReader implements BaseReader {
...
+      fs,
+      defaultReader,
+      fileExtToReader,
+      overrideReader,
+    };
 
-        let reader: BaseReader;
+    const limit = pLimit(numWorkers);
+    const workerPromises = filePathQueue.map((filePath) =>
+      limit(() => this.processFiles(filePath, processFilesParams)),
```
Comment (marcusschiesser) [inline packages/core/src/readers/SimpleDirectoryReader.edge.ts:89]: isn't it processing just one file?
```suggestion
      limit(() => this.processFile(filePath, processFilesParams)),
```

File: examples/readers/package.json [hunk 5]
```diff
@@ -11,7 +11,7 @@
     "start:pdf": "node --import tsx ./src/pdf.ts",
     "start:llamaparse": "node --import tsx ./src/llamaparse.ts",
     "start:notion": "node --import tsx ./src/notion.ts",
-    "start:llamaparse2": "node --import tsx ./src/llamaparse_2.ts"
+    "start:llamaparse2": "node --import tsx ./src/simple-directory-reader-with-llamaparse.ts"
```
Comment (marcusschiesser) [inline examples/readers/package.json:14]:
```suggestion
    "start:llamaparse-dir": "node --import tsx ./src/simple-directory-reader-with-llamaparse.ts"
```

File: .changeset/stale-dots-search.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+"llamaindex": patch
+---
+
+add p-limit for enhanced concurrency managment
```
Comment (marcusschiesser) [inline .changeset/stale-dots-search.md:5]:
```suggestion
add concurrency managment for SimpleDirectoryReader
```

**Review by marcusschiesser — COMMENTED** [review]
@MightBeAScam looks great! just a few notes and it needs to be rebased

**Review by himself65 — CHANGES_REQUESTED** [review]
This account has been deleted, closing

**Comment by himself65** [comment]
image

**Comment by KindOfAScam** [comment]
@himself65 sorry my bad, realized 2 accounts are stupid, so I deleted the newer one..

But I messed up rebasing anyway, so let me try again and I'll open a new PR
